File: examples/python/connect.py
"""
connect.py — Establish a WebSocket connection to an iMRP chassis via ROPA.

ROPA uses WebSocket on port 9090, JSON message format, full-duplex.
This example connects and prints every incoming message.

Requirements: pip install websocket-client
"""

import websocket
import json
import sys

ROBOT_IP = "192.168.1.100"  # Replace with your chassis IP
ROPA_PORT = 9090


def on_message(ws, message):
    """Handle every message pushed by the robot."""
    try:
        data = json.loads(message)
        op = data.get("op", "unknown")
        if op == "publish":
            print(f"[TOPIC] {data.get('topic')} → {json.dumps(data.get('msg', {}), indent=2)[:200]}")
        elif op == "service_response":
            print(f"[SERVICE] result={data.get('result')}, values={data.get('values', {})}")
        else:
            print(f"[{op}] {json.dumps(data)[:200]}")
    except json.JSONDecodeError:
        print(f"[RAW] {message[:200]}")


def on_error(ws, error):
    print(f"❌ WebSocket error: {error}")


def on_close(ws, close_status_code, close_msg):
    print(f"🔌 Connection closed (code={close_status_code}, msg={close_msg})")


def on_open(ws):
    print("✅ Connected to iMRP chassis via ROPA!")
    print(f"   WebSocket: ws://{ROBOT_IP}:{ROPA_PORT}")
    print("   Listening for all incoming messages...")

    # Subscribe to robot status as a basic heartbeat check
    ws.send(json.dumps({
        "op": "subscribe",
        "topic": "/robot_status",
        "type": "robot_status"
    }))
    print("   → Subscribed to /robot_status")


def main():
    url = f"ws://{ROBOT_IP}:{ROPA_PORT}"
    print(f"Connecting to {url} ...")

    ws = websocket.WebSocketApp(
        url,
        on_open=on_open,
        on_message=on_message,
        on_error=on_error,
        on_close=on_close
    )

    try:
        ws.run_forever()
    except KeyboardInterrupt:
        print("\nDisconnecting...")
        ws.close()


if __name__ == "__main__":
    if len(sys.argv) > 1:
        ROBOT_IP = sys.argv[1]
    main()

File: examples/javascript/connect.js
/**
 * connect.js — Connect to an iMRP chassis via ROPA WebSocket.
 *
 * ROPA: WebSocket port 9090, JSON format, full-duplex.
 * This example connects and logs all incoming messages.
 *
 * Run: node connect.js [ROBOT_IP]
 */

const WebSocket = require("ws");

const ROBOT_IP = process.argv[2] || "192.168.1.100";
const ROPA_PORT = 9090;
const url = `ws://${ROBOT_IP}:${ROPA_PORT}`;

console.log(`Connecting to ${url} ...`);

const ws = new WebSocket(url);

ws.on("open", () => {
  console.log("✅ Connected to iMRP chassis via ROPA!");
  console.log(`   WebSocket: ${url}`);
  console.log("   Listening for all incoming messages...");

  // Subscribe to robot status as a basic heartbeat
  ws.send(JSON.stringify({
    op: "subscribe",
    topic: "/robot_status",
    type: "robot_status"
  }));
  console.log("   → Subscribed to /robot_status");
});

ws.on("message", (raw) => {
  try {
    const data = JSON.parse(raw);
    const op = data.op || "unknown";

    if (op === "publish") {
      const topic = data.topic || "";
      const msg = data.msg || {};
      console.log(`[TOPIC] ${topic} → ${JSON.stringify(msg).substring(0, 200)}`);
    } else if (op === "service_response") {
      console.log(`[SERVICE] result=${data.result}, values=${JSON.stringify(data.values || {}).substring(0, 200)}`);
    } else {
      console.log(`[${op}] ${JSON.stringify(data).substring(0, 200)}`);
    }
  } catch (e) {
    console.log(`[RAW] ${raw.toString().substring(0, 200)}`);
  }
});

ws.on("error", (err) => {
  console.log(`❌ WebSocket error: ${err.message}`);
});

ws.on("close", (code, reason) => {
  console.log(`🔌 Connection closed (code=${code}, reason=${reason || "none"})`);
});

// Keep the process alive
process.on("SIGINT", () => {
  console.log("\nDisconnecting...");
  ws.close();
  process.exit(0);
});
